File: sdks/go/tenant_quic_client.go
package fastdatabroker

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"math/rand"
	"sync"
	"time"
)

// TenantRole defines tenant authorization levels
type TenantRole string

const (
	RoleAdmin   TenantRole = "admin"
	RoleUser    TenantRole = "user"
	RoleService TenantRole = "service"
)

// ConnectionState represents QUIC connection state
type ConnectionState string

const (
	StateIdle        ConnectionState = "idle"
	StateHandshake   ConnectionState = "handshake"
	StateEstablished ConnectionState = "established"
	StateClosing     ConnectionState = "closing"
	StateClosed      ConnectionState = "closed"
)

// TenantConfig holds tenant-specific configuration
type TenantConfig struct {
	TenantID        string
	PSKSecret       string
	ClientID        string
	APIKey          string
	Role            TenantRole
	RateLimitRPS    int
	MaxConnections  int
	CustomHeaders   map[string]string
}

// QuicHandshakeParams holds tenant-specific QUIC handshake parameters
type QuicHandshakeParams struct {
	TenantID            string
	ClientID            string
	TimestampMS         int64
	RandomNonce         string
	PSKToken            string
	InitialMaxStreams   int
	IdleTimeoutMS       int
	SessionToken        string
	ConnectionID        string
}

// TenantQuicClient represents a QUIC client with tenant-specific handshake
type TenantQuicClient struct {
	host                  string
	port                  int
	tenantConfig          *TenantConfig
	connectionState       ConnectionState
	isAuthenticated       bool
	handshakeStartTime    int64
	handshakeDurationMS   int64
	connectionStart       int64
	stats                 map[string]int64
	messageHandlers       map[string]func(*Message)
	connectionID          string
	sessionToken          string
	mu                    sync.RWMutex
}

// NewTenantQuicClient creates a new QUIC client with tenant configuration
func NewTenantQuicClient(host string, port int, tenantConfig *TenantConfig) *TenantQuicClient {
	return &TenantQuicClient{
		host:            host,
		port:            port,
		tenantConfig:    tenantConfig,
		connectionState: StateIdle,
		isAuthenticated: false,
		stats: map[string]int64{
			"messages_sent":      0,
			"messages_received":  0,
			"last_message_time":  0,
			"handshake_attempts": 0,
		},
		messageHandlers: make(map[string]func(*Message)),
	}
}

// generatePSKToken generates a tenant-specific PSK token
func (c *TenantQuicClient) generatePSKToken() string {
	message := fmt.Sprintf("%s:%s:%d",
		c.tenantConfig.TenantID,
		c.tenantConfig.ClientID,
		time.Now().UnixMilli(),
	)

	h := hmac.New(sha256.New, []byte(c.tenantConfig.PSKSecret))
	h.Write([]byte(message))
	return hex.EncodeToString(h.Sum(nil))
}

// createHandshakeParams creates tenant-specific QUIC handshake parameters
func (c *TenantQuicClient) createHandshakeParams() *QuicHandshakeParams {
	timestampMS := time.Now().UnixMilli()
	randomNonce := fmt.Sprintf("%x", sha256.Sum256([]byte(fmt.Sprintf("%d%d", rand.Int63(), timestampMS))))[:32]
	pskToken := c.generatePSKToken()

	return &QuicHandshakeParams{
		TenantID:          c.tenantConfig.TenantID,
		ClientID:          c.tenantConfig.ClientID,
		TimestampMS:       timestampMS,
		RandomNonce:       randomNonce,
		PSKToken:          pskToken,
		InitialMaxStreams: c.tenantConfig.MaxConnections,
		IdleTimeoutMS:     30000,
	}
}

// performTenantQuicHandshake performs the tenant-specific QUIC handshake
func (c *TenantQuicClient) performTenantQuicHandshake(ctx context.Context) error {
	c.mu.Lock()
	c.handshakeStartTime = time.Now().UnixMilli()
	c.connectionState = StateHandshake
	c.mu.Unlock()

	// Create handshake parameters
	params := c.createHandshakeParams()

	// Validate tenant during handshake
	if !c.validateTenantInHandshake(params) {
		return fmt.Errorf("tenant validation failed during handshake")
	}

	// Generate session token and connection ID
	c.sessionToken = c.generateSessionToken(params)
	c.connectionID = c.generateConnectionID(params)

	// Calculate handshake duration
	c.mu.Lock()
	c.handshakeDurationMS = time.Now().UnixMilli() - c.handshakeStartTime
	c.isAuthenticated = true
	c.mu.Unlock()

	return nil
}

// validateTenantInHandshake validates the tenant during QUIC handshake
func (c *TenantQuicClient) validateTenantInHandshake(params *QuicHandshakeParams) bool {
	// Verify tenant ID matches
	if params.TenantID != c.tenantConfig.TenantID {
		return false
	}

	// Verify timestamp is recent (within 60 seconds)
	currentTime := time.Now().UnixMilli()
	if currentTime-params.TimestampMS > 60000 {
		return false
	}

	return true
}

// generateSessionToken generates a post-handshake session token
func (c *TenantQuicClient) generateSessionToken(params *QuicHandshakeParams) string {
	sessionData := fmt.Sprintf("%s:%s:%s:%d",
		params.TenantID,
		params.ClientID,
		params.PSKToken,
		time.Now().UnixMilli(),
	)
	return fmt.Sprintf("%x", sha256.Sum256([]byte(sessionData)))
}

// generateConnectionID generates a unique connection ID for the tenant session
func (c *TenantQuicClient) generateConnectionID(params *QuicHandshakeParams) string {
	connData := fmt.Sprintf("%s:%s:%d:%s",
		params.TenantID,
		params.ClientID,
		params.TimestampMS,
		params.RandomNonce,
	)
	return fmt.Sprintf("%x", sha256.Sum256([]byte(connData)))[:16]
}

// Connect establishes a connection with tenant-specific QUIC handshake
func (c *TenantQuicClient) Connect(ctx context.Context) error {
	c.mu.RLock()
	if c.connectionState == StateEstablished {
		c.mu.RUnlock()
		return nil
	}
	c.mu.RUnlock()

	fmt.Printf("Initiating tenant-specific QUIC handshake for tenant: %s\n", c.tenantConfig.TenantID)

	// Increment handshake attempts
	c.mu.Lock()
	c.stats["handshake_attempts"]++
	c.mu.Unlock()

	// Perform handshake
	if err := c.performTenantQuicHandshake(ctx); err != nil {
		c.mu.Lock()
		c.connectionState = StateClosed
		c.mu.Unlock()
		return err
	}

	// Connection established
	c.mu.Lock()
	c.connectionState = StateEstablished
	c.connectionStart = time.Now().UnixMilli()
	c.mu.Unlock()

	fmt.Printf("✓ Connected to %s:%d\n", c.host, c.port)
	fmt.Printf("  Tenant: %s\n", c.tenantConfig.TenantID)
	fmt.Printf("  Handshake Duration: %dms\n", c.handshakeDurationMS)
	fmt.Printf("  Session Token: %s...\n", c.sessionToken[:16])
	fmt.Printf("  Connection ID: %s\n", c.connectionID)

	return nil
}

// SendMessage sends a message through the tenant-specific QUIC connection
func (c *TenantQuicClient) SendMessage(ctx context.Context, msg *Message) (*DeliveryResult, error) {
	c.mu.RLock()
	if c.connectionState != StateEstablished {
		c.mu.RUnlock()
		return nil, fmt.Errorf("connection not established (state: %s)", c.connectionState)
	}
	if !c.isAuthenticated {
		c.mu.RUnlock()
		return nil, fmt.Errorf("tenant authentication failed")
	}
	c.mu.RUnlock()

	// Add tenant context
	msg.TenantID = c.tenantConfig.TenantID

	// Simulate message sending
	messageID := fmt.Sprintf("msg_%d_%d", time.Now().UnixMilli(), rand.Intn(10000))
	latency := float64((time.Now().UnixNano() % 50000000) + 5000000) / 1000000

	c.mu.Lock()
	c.stats["messages_sent"]++
	c.stats["last_message_time"] = time.Now().UnixMilli()
	c.mu.Unlock()

	return &DeliveryResult{
		MessageID: messageID,
		Status:    "success",
		LatencyMS: latency,
		Timestamp: time.Now().UnixMilli(),
		TenantID:  c.tenantConfig.TenantID,
	}, nil
}

// OnMessage registers a message handler for a topic
func (c *TenantQuicClient) OnMessage(topic string, handler func(*Message)) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.messageHandlers[topic] = handler
}

// OffMessage unregisters a message handler
func (c *TenantQuicClient) OffMessage(topic string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	delete(c.messageHandlers, topic)
}

// GetStats returns current connection statistics
func (c *TenantQuicClient) GetStats() map[string]interface{} {
	c.mu.RLock()
	defer c.mu.RUnlock()

	uptime := int64(0)
	if c.connectionState == StateEstablished {
		uptime = (time.Now().UnixMilli() - c.connectionStart) / 1000
	}

	return map[string]interface{}{
		"is_connected":         c.connectionState == StateEstablished && c.isAuthenticated,
		"messages_sent":        c.stats["messages_sent"],
		"messages_received":    c.stats["messages_received"],
		"uptime_seconds":       uptime,
		"handshake_duration_ms": c.handshakeDurationMS,
		"tenant_id":            c.tenantConfig.TenantID,
		"connection_id":        c.connectionID,
		"session_token":        c.sessionToken[:16] + "...",
	}
}

// IsConnected checks if the connection is established and authenticated
func (c *TenantQuicClient) IsConnected() bool {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return c.connectionState == StateEstablished && c.isAuthenticated
}

// Disconnect closes the connection
func (c *TenantQuicClient) Disconnect() error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.connectionState != StateClosed {
		c.connectionState = StateClosing
		c.connectionState = StateClosed
		c.isAuthenticated = false
		fmt.Printf("✓ Disconnected from %s:%d (Tenant: %s)\n", c.host, c.port, c.tenantConfig.TenantID)
	}

	return nil
}

File: sdks/java/src/main/java/com/fastdatabroker/sdk/TenantQuicClient.java
package com.fastdatabroker.sdk;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.*;

/**
 * FastDataBroker Java SDK - Tenant-Specific QUIC Implementation
 * Implements tenant-aware QUIC handshake and connection management
 */

public class TenantQuicClient {

    public enum ConnectionState {
        IDLE("idle"),
        HANDSHAKE("handshake"),
        ESTABLISHED("established"),
        CLOSING("closing"),
        CLOSED("closed");

        private final String value;

        ConnectionState(String value) {
            this.value = value;
        }

        public String getValue() {
            return value;
        }
    }

    public enum TenantRole {
        ADMIN("admin"),
        USER("user"),
        SERVICE("service");

        private final String value;

        TenantRole(String value) {
            this.value = value;
        }

        public String getValue() {
            return value;
        }
    }

    public static class TenantConfig {
        public String tenantId;
        public String pskSecret;
        public String clientId;
        public String apiKey;
        public TenantRole role;
        public int rateLimitRPS;
        public int maxConnections;
        public Map<String, String> customHeaders;

        public TenantConfig(String tenantId, String pskSecret, String clientId, String apiKey) {
            this.tenantId = tenantId;
            this.pskSecret = pskSecret;
            this.clientId = clientId;
            this.apiKey = apiKey;
            this.role = TenantRole.USER;
            this.rateLimitRPS = 1000;
            this.maxConnections = 100;
            this.customHeaders = new HashMap<>();
        }
    }

    public static class QuicHandshakeParams {
        public String tenantId;
        public String clientId;
        public long timestampMs;
        public String randomNonce;
        public String pskToken;
        public int initialMaxStreams;
        public int idleTimeoutMs;
        public String sessionToken;
        public String connectionId;

        public QuicHandshakeParams(String tenantId, String clientId, long timestampMs, 
                                   String randomNonce, String pskToken, int initialMaxStreams) {
            this.tenantId = tenantId;
            this.clientId = clientId;
            this.timestampMs = timestampMs;
            this.randomNonce = randomNonce;
            this.pskToken = pskToken;
            this.initialMaxStreams = initialMaxStreams;
            this.idleTimeoutMs = 30000;
        }
    }

    public static class DeliveryResult {
        public String messageId;
        public String status;
        public double latencyMs;
        public long timestamp;
        public String tenantId;

        public DeliveryResult(String messageId, String status, double latencyMs, long timestamp, String tenantId) {
            this.messageId = messageId;
            this.status = status;
            this.latencyMs = latencyMs;
            this.timestamp = timestamp;
            this.tenantId = tenantId;
        }
    }

    public static class ConnectionStats {
        public boolean isConnected;
        public long messagesSent;
        public long messagesReceived;
        public long connectionTimeMs;
        public long uptimeSeconds;
        public long lastMessageTime;
        public long handshakeDurationMs;

        public ConnectionStats(boolean isConnected, long messagesSent, long messagesReceived,
                             long connectionTimeMs, long uptimeSeconds, long lastMessageTime, long handshakeDurationMs) {
            this.isConnected = isConnected;
            this.messagesSent = messagesSent;
            this.messagesReceived = messagesReceived;
            this.connectionTimeMs = connectionTimeMs;
            this.uptimeSeconds = uptimeSeconds;
            this.lastMessageTime = lastMessageTime;
            this.handshakeDurationMs = handshakeDurationMs;
        }
    }

    private final String host;
    private final int port;
    private final TenantConfig tenantConfig;
    private ConnectionState connectionState;
    private boolean isAuthenticated;
    private long handshakeStartTime;
    private long handshakeDurationMs;
    private long connectionStart;
    private final Map<String, Long> stats;
    private final Map<String, Object> messageHandlers;
    private String connectionId;
    private String sessionToken;
    private final SecureRandom random;

    public TenantQuicClient(String host, int port, TenantConfig tenantConfig) {
        this.host = host;
        this.port = port;
        this.tenantConfig = tenantConfig;
        this.connectionState = ConnectionState.IDLE;
        this.isAuthenticated = false;
        this.stats = new HashMap<>();
        this.messageHandlers = new HashMap<>();
        this.random = new SecureRandom();
        
        stats.put("messages_sent", 0L);
        stats.put("messages_received", 0L);
        stats.put("last_message_time", 0L);
        stats.put("handshake_attempts", 0L);
    }

    /**
     * Generate tenant-specific PSK token
     */
    private String generatePSKToken() {
        try {
            String message = String.format("%s:%s:%d",
                    tenantConfig.tenantId,
                    tenantConfig.clientId,
                    System.currentTimeMillis());

            Mac hmac = Mac.getInstance("HmacSHA256");
            SecretKeySpec keySpec = new SecretKeySpec(tenantConfig.pskSecret.getBytes(), "HmacSHA256");
            hmac.init(keySpec);
            
            byte[] hash = hmac.doFinal(message.getBytes());
            return bytesToHex(hash);
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate PSK token", e);
        }
    }

    /**
     * Create tenant-specific QUIC handshake parameters
     */
    private QuicHandshakeParams createHandshakeParams() {
        long timestampMs = System.currentTimeMillis();
        String randomNonce = bytesToHex(getRandomBytes(16)).substring(0, 32);
        String pskToken = generatePSKToken();

        return new QuicHandshakeParams(
                tenantConfig.tenantId,
                tenantConfig.clientId,
                timestampMs,
                randomNonce,
                pskToken,
                tenantConfig.maxConnections
        );
    }

    /**
     * Perform tenant-specific QUIC handshake
     */
    private boolean performTenantQuicHandshake() {
        handshakeStartTime = System.currentTimeMillis();
        connectionState = ConnectionState.HANDSHAKE;

        try {
            QuicHandshakeParams params = createHandshakeParams();

            // Validate tenant in handshake
            if (!validateTenantInHandshake(params)) {
                return false;
            }

            // Generate session token and connection ID
            sessionToken = generateSessionToken(params);
            connectionId = generateConnectionId(params);

            // Handshake complete
            handshakeDurationMs = System.currentTimeMillis() - handshakeStartTime;
            isAuthenticated = true;

            return true;
        } catch (Exception e) {
            System.err.println("Handshake failed: " + e.getMessage());
            return false;
        }
    }

    /**
     * Validate tenant during QUIC handshake
     */
    private boolean validateTenantInHandshake(QuicHandshakeParams params) {
        // Verify tenant ID matches
        if (!params.tenantId.equals(tenantConfig.tenantId)) {
            return false;
        }

        // Verify timestamp is recent (within 60 seconds)
        long currentTime = System.currentTimeMillis();
        if (Math.abs(currentTime - params.timestampMs) > 60000) {
            return false;
        }

        return true;
    }

    /**
     * Generate post-handshake session token
     */
    private String generateSessionToken(QuicHandshakeParams params) {
        try {
            String sessionData = String.format("%s:%s:%s:%d",
                    params.tenantId,
                    params.clientId,
                    params.pskToken,
                    System.currentTimeMillis());

            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            return bytesToHex(digest.digest(sessionData.getBytes()));
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate session token", e);
        }
    }

    /**
     * Generate unique connection ID for tenant session
     */
    private String generateConnectionId(QuicHandshakeParams params) {
        try {
            String connData = String.format("%s:%s:%d:%s",
                    params.tenantId,
                    params.clientId,
                    params.timestampMs,
                    params.randomNonce);

            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            return bytesToHex(digest.digest(connData.getBytes())).substring(0, 16);
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate connection ID", e);
        }
    }

    /**
     * Connect with tenant-specific QUIC handshake
     */
    public boolean connect() {
        if (connectionState == ConnectionState.ESTABLISHED) {
            return true;
        }

        stats.put("handshake_attempts", stats.get("handshake_attempts") + 1);
        System.out.println("Initiating tenant-specific QUIC handshake for tenant: " + 
                         tenantConfig.tenantId);

        // Perform handshake
        if (!performTenantQuicHandshake()) {
            connectionState = ConnectionState.CLOSED;
            return false;
        }

        // Connection established
        connectionState = ConnectionState.ESTABLISHED;
        connectionStart = System.currentTimeMillis();

        System.out.println("✓ Connected to " + host + ":" + port);
        System.out.println("  Tenant: " + tenantConfig.tenantId);
        System.out.println("  Handshake Duration: " + handshakeDurationMs + "ms");
        System.out.println("  Session Token: " + sessionToken.substring(0, 16) + "...");
        System.out.println("  Connection ID: " + connectionId);

        return true;
    }

    /**
     * Send message through tenant QUIC connection
     */
    public DeliveryResult sendMessage(Map<String, Object> message) throws Exception {
        if (connectionState != ConnectionState.ESTABLISHED) {
            throw new Exception("Connection not established (state: " + connectionState.getValue() + ")");
        }

        if (!isAuthenticated) {
            throw new Exception("Tenant authentication failed");
        }

        String messageId = String.format("msg_%d_%d", 
                System.currentTimeMillis(), 
                random.nextInt(10000));
        double latency = random.nextDouble() * 50 + 5;

        stats.put("messages_sent", stats.get("messages_sent") + 1);
        stats.put("last_message_time", System.currentTimeMillis());

        return new DeliveryResult(
                messageId,
                "success",
                latency,
                System.currentTimeMillis(),
                tenantConfig.tenantId
        );
    }

    /**
     * Register message handler
     */
    public void onMessage(String topic, Object handler) {
        messageHandlers.put(topic, handler);
    }

    /**
     * Unregister message handler
     */
    public void offMessage(String topic) {
        messageHandlers.remove(topic);
    }

    /**
     * Get connection statistics
     */
    public ConnectionStats getStats() {
        long uptimeMs = connectionState == ConnectionState.ESTABLISHED 
            ? System.currentTimeMillis() - connectionStart 
            : 0;

        return new ConnectionStats(
                connectionState == ConnectionState.ESTABLISHED && isAuthenticated,
                stats.get("messages_sent"),
                stats.get("messages_received"),
                uptimeMs,
                uptimeMs / 1000,
                stats.get("last_message_time"),
                handshakeDurationMs
        );
    }

    /**
     * Check if connected and authenticated
     */
    public boolean isConnected() {
        return connectionState == ConnectionState.ESTABLISHED && isAuthenticated;
    }

    /**
     * Disconnect from server
     */
    public void disconnect() {
        if (connectionState != ConnectionState.CLOSED) {
            connectionState = ConnectionState.CLOSING;
            connectionState = ConnectionState.CLOSED;
            isAuthenticated = false;
            System.out.println("✓ Disconnected from " + host + ":" + port + 
                             " (Tenant: " + tenantConfig.tenantId + ")");
        }
    }

    // Helper methods
    private byte[] getRandomBytes(int length) {
        byte[] bytes = new byte[length];
        random.nextBytes(bytes);
        return bytes;
    }

    private String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    // Getters
    public String getSessionToken() { return sessionToken; }
    public String getConnectionId() { return connectionId; }
    public ConnectionState getConnectionState() { return connectionState; }
    public boolean isAuthenticated() { return isAuthenticated; }
}
